From da55dc61e8818e84519e4d64005b2447dbee311b Mon Sep 17 00:00:00 2001
From: Wanchang Ryu <wanchang.ryu@lge.com>
Date: Thu, 3 Sep 2020 10:36:19 +0900
Subject: [PATCH] waitForStop API is added for cleanup message handler thread

uMediaClient creates a thread for dispatching message from uMediaServer
and it calls derived virtual method when dispatching message.
There is a race condition between destructing and dispatching message.
Event dispatch still can be happened even after called derived class's
destructor since the thread for message dispatching is cleaned up on
the destructor of the base class.

This CL added |waitForStop| API to cleanup message thread before exiting
destructor of derived class.

---
 include/public/uMediaClient.h     | 1 +
 src/media_client/uMediaClient.cpp | 5 +++++
 2 files changed, 6 insertions(+)

diff --git a/include/public/uMediaClient.h b/include/public/uMediaClient.h
index ca79789..04a1097 100644
--- a/include/public/uMediaClient.h
+++ b/include/public/uMediaClient.h
@@ -414,6 +414,7 @@ public :
 
 	void run();
 	void stop();
+  void waitForStop();
 
 protected :
 	UMSConnector *connection;
diff --git a/src/media_client/uMediaClient.cpp b/src/media_client/uMediaClient.cpp
index 0794bfc..96a9d90 100644
--- a/src/media_client/uMediaClient.cpp
+++ b/src/media_client/uMediaClient.cpp
@@ -1441,6 +1441,11 @@ void uMediaClient::stop()
   connection->stop();
 }
 
+void uMediaClient::waitForStop() {
+  stop();
+  pthread_join(message_thread, NULL);
+}
+
 // -----------------------------------------------
 // --- utility functions to create JSON arguments
 
